File: hl2da_unity/Assets/Scripts/hl2da_framebuffer.cs

using System;

public class hl2da_framebuffer
{
    public enum STATUS
    {
        DISCARDED = -1,
        OK = 0,
        WAIT = 1,
    }

    public enum TIME_PREFERENCE
    {
        PAST = -1,
        NEAREST = 0,
        FUTURE = 1,
    }

    private const int BUFFER_COUNT = 4;

    private IntPtr frame;

    private STATUS status;
    private hl2da_api.SENSOR_ID id;
    private ulong timestamp;
    private int framestamp;
    private int valid;

    private IntPtr[] buffer;
    private int[] length;

    public hl2da_framebuffer()
    {
        Reset();
    }

    ~hl2da_framebuffer()
    {
        Destroy();
    }

    public STATUS Status { get { return status; } }

    public hl2da_api.SENSOR_ID Id { get { return id; } }

    public ulong Timestamp { get { return timestamp; } }

    public int Framestamp { get { return framestamp; } }

    public int Valid { get { return valid; } }

    public IntPtr Buffer(uint index)
    {
        if (index >= BUFFER_COUNT) { return IntPtr.Zero; }
        return buffer[index];
    }

    public int Length(uint index)
    {
        if (index >= BUFFER_COUNT) { return 0; }
        return length[index];
    }

    private void Reset()
    {
        frame = IntPtr.Zero;
        status = STATUS.DISCARDED;
        id = (hl2da_api.SENSOR_ID)(-1);
        timestamp = 0;
        framestamp = 0;
        valid = 0;
        buffer = new IntPtr[4];
        length = new int[4];
    }

    public void Destroy()
    {
        if (frame == IntPtr.Zero) { return; }
        hl2da_api.Release((int)id, frame);
        Reset();
    }

    private static void Extract(hl2da_framebuffer fb)
    {
        hl2da_api.Extract((int)fb.id, fb.frame, ref fb.valid, fb.buffer, fb.length);
    }

    public static hl2da_framebuffer GetFrame(hl2da_api.SENSOR_ID id, int framestamp)
    {
        hl2da_framebuffer fb = new hl2da_framebuffer();
        fb.id = id;
        fb.status = (STATUS)hl2da_api.GetByFramestamp((int)id, framestamp, ref fb.frame, ref fb.timestamp, ref fb.framestamp);
        if (fb.frame != IntPtr.Zero) { Extract(fb); }
        return fb;
    }

    public static hl2da_framebuffer GetFrame(hl2da_api.SENSOR_ID id, ulong timestamp, TIME_PREFERENCE time_preference, bool tiebreak_right)
    {
        hl2da_framebuffer fb = new hl2da_framebuffer();
        fb.id = id;
        fb.status = (STATUS)hl2da_api.GetByTimestamp((int)id, timestamp, (int)time_preference, tiebreak_right ? 1 : 0, ref fb.frame, ref fb.timestamp, ref fb.framestamp);
        if (fb.frame != IntPtr.Zero) { Extract(fb); }
        return fb;
    }
}

File: hl2da_unreal/Plugins/hl2da/Source/hl2da/Private/hl2da_framebuffer.cpp
// Fill out your copyright notice in the Description page of Project Settings.


#include "hl2da_framebuffer.h"
#include "hl2da_api.h"

hl2da_framebuffer::hl2da_framebuffer()
{
    Reset();
}

hl2da_framebuffer::~hl2da_framebuffer()
{
    Destroy();
}

int32_t hl2da_framebuffer::Status()
{
    return status;
}

int32_t hl2da_framebuffer::Id()
{
    return id;
}

uint64_t hl2da_framebuffer::Timestamp()
{
    return timestamp;
}

int32_t hl2da_framebuffer::Framestamp()
{
    return framestamp;
}

int32_t hl2da_framebuffer::Valid()
{
    return valid;
}

void const* hl2da_framebuffer::Buffer(uint32_t index)
{
    if (index >= (sizeof(buffer) / sizeof(void*))) { return nullptr; }
    return buffer[index];
}

int32_t hl2da_framebuffer::Length(uint32_t index)
{
    if (index >= (sizeof(buffer) / sizeof(void*))) { return 0; }
    return length[index];
}

void hl2da_framebuffer::Reset()
{
    frame = nullptr;
    status = -1;
    id = -1;
    timestamp = 0;
    framestamp = 0;
    valid = 0;
    memset(buffer, 0, sizeof(buffer));
    memset(length, 0, sizeof(length));
}

void hl2da_framebuffer::Destroy()
{
    if (!frame) { return; }
    hl2da_api::Release(id, frame);
    Reset();
}

void hl2da_framebuffer::Extract(hl2da_framebuffer& fb)
{
    hl2da_api::Extract(fb.id, fb.frame, &fb.valid, fb.buffer, fb.length);
}

std::shared_ptr<hl2da_framebuffer> hl2da_framebuffer::GetFrame(int id, int framestamp)
{
    std::shared_ptr<hl2da_framebuffer> fb = std::make_shared<hl2da_framebuffer>();
    fb->id = id;
    fb->status = hl2da_api::GetByFramestamp(id, framestamp, &fb->frame, &fb->timestamp, &fb->framestamp);
    if (fb->frame) { Extract(*fb); }
    return fb;
}

std::shared_ptr<hl2da_framebuffer> hl2da_framebuffer::GetFrame(int id, uint64_t timestamp, int time_preference, bool tiebreak_right)
{
    std::shared_ptr<hl2da_framebuffer> fb = std::make_shared<hl2da_framebuffer>();
    fb->id = id;
    fb->status = hl2da_api::GetByTimestamp(id, timestamp, time_preference, tiebreak_right, &fb->frame, &fb->timestamp, &fb->framestamp);
    if (fb->frame) { Extract(*fb); }
    return fb;
}
